internal/pagefile.go

```diff
@@ -136,73 +136,179 @@ func ReadIncrementalFile(filePath string, fileSize int64, lsn uint64, deltaBitma
 	return pageReader, incrementSize, nil
 }
 
-// ApplyFileIncrement changes pages according to supplied change map file
-func ApplyFileIncrement(fileName string, increment io.Reader, createNewIncrementalFiles bool) error {
-	tracelog.DebugLogger.Printf("Incrementing %s\n", fileName)
-	err := ReadIncrementFileHeader(increment)
+// RestoreMissingPages restores missing pages (zero blocks)
+// of local file with their base backup version
+func RestoreMissingPages(base io.Reader, file *os.File) error {
+	tracelog.DebugLogger.Printf("Restoring missing pages from base backup: %s\n", file.Name())
+	defer file.Sync()
+
+	filePageCount, err := getPageCount(file)
 	if err != nil {
 		return err
 	}
+	for i := int64(0); i < filePageCount; i++ {
+		err = writePage(file, i, base, false)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+	}
 
-	var fileSize uint64
-	var diffBlockCount uint32
-	err = parsingutil.ParseMultipleFieldsFromReader([]parsingutil.FieldToParse{
-		{Field: &fileSize, Name: "fileSize"},
-		{Field: &diffBlockCount, Name: "diffBlockCount"},
-	}, increment)
-	if err != nil {
-		return err
+	all, _ := base.Read(make([]byte, 1))
+	if all > 0 {
+		tracelog.DebugLogger.Printf("Skipping pages after end of the local file %s, " +
+			"possibly the pagefile was truncated.\n", file.Name())
 	}
+	return nil
+}
 
-	diffMap := make([]byte, diffBlockCount*sizeofInt32)
+// CreateFileFromIncrement writes the pages from the increment to local file
+// and write empty blocks in place of pages which are not present in the increment
+func CreateFileFromIncrement(increment io.Reader, file *os.File) error {
+	tracelog.DebugLogger.Printf("Generating file from increment %s\n", file.Name())
+	defer file.Sync()
 
-	_, err = io.ReadFull(increment, diffMap)
+	fileSize, diffBlockCount, diffMap, err := getIncrementHeaderFields(increment)
 	if err != nil {
 		return err
 	}
 
-	openFlags := os.O_RDWR
-	if createNewIncrementalFiles {
-		openFlags = openFlags | os.O_CREATE
+	// set represents all block numbers with non-empty pages
+	deltaBlockNumbers := make(map[uint32]bool, diffBlockCount)
+	for i := uint32(0); i < diffBlockCount; i++ {
+		blockNo := binary.LittleEndian.Uint32(diffMap[i*sizeofInt32 : (i+1)*sizeofInt32])
+		deltaBlockNumbers[blockNo] = true
 	}
-
-	file, err := os.OpenFile(fileName, openFlags, 0666)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return errors.Wrap(err, "incremented file should always exist")
+	pageCount := uint32(fileSize / uint64(DatabasePageSize))
```

Comment:
Ok, will do

I left it there because I thought that the page count will never overflow uint32.
The only way to overflow uint32 here (assuming that page size is equal to 8KB) is to use a 35TB pagefile, but postgres default is 1GB.

Reply:
I changed it to int64 in 74de388d16b637a0c88cbeeebbb6094877536373